Copter: Add support for pausing and continuing auto missions

mustafa-gokce opened the pull request:
This pull request adds support MAV_CMD_DO_PAUSE_CONTINUE command in AUTO mode for Copter. When copter receives pause command `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0 0 0` it loiters, and when it receives continue command `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 1 0 0 0 0 0 0` it resumes the mission.

amilcarlucas commented:
Thanks, it is much better now

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -695,6 +695,32 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_int_packet(const mavlink_command_i
     case MAV_CMD_DO_REPOSITION:
         return handle_command_int_do_reposition(packet);
+
+    // pause or resume an auto mission
+    case MAV_CMD_DO_PAUSE_CONTINUE:
```
The code in here should be factored out into a `handle_command_pause_continue` method, and also be available from `handle_command_long_packet`.

peterbarker reviewed (changes requested):
I quite like this, and I think we should merge this functionality without expanding on it.

However, I will point out that the description of the mavlink command states that it should work in any GPS-controlled position mode, and I think it would make sense to honour that in at least `GUIDED` mode, too.

This command would also be appropriate in all of our other vehicles, but I can't see that they could share any code at this point.

rmackay9 commented:
This looks pretty OK to me.  This has been tested in SITL of course?

mustafa-gokce commented:
@peterbarker now MAV_CMD_PAUSE_CONTINUE commands can be sent with both as COMMAND_LONG and COMMAND_INT commands.
I testes this in SITL by sending commands like:
```
AUTO> command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0 0 0
AUTO> command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 1 0 0 0 0 0 0
AUTO> long MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0
AUTO> long MAV_CMD_DO_PAUSE_CONTINUE 1 0 0 0 0 0 0
```
It is working properly on SITL.

peterbarker commented:
@mustafa-gokce thanks for fixing this up.

We don't allow merge commits into ArduPilot, however - please do a rebase on top of master to remove it.

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -963,11 +968,43 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
     }
 
+    // pause or resume an auto mission
+    case MAV_CMD_DO_PAUSE_CONTINUE:
```
```suggestion
    case MAV_CMD_DO_PAUSE_CONTINUE: {
```

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -963,11 +968,43 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
+        mavlink_command_int_t packet_int;
+        GCS_MAVLINK_Copter::convert_COMMAND_LONG_to_COMMAND_INT(packet, packet_int);
+        return handle_command_pause_continue(packet_int);
+
```
```suggestion
    }
```

peterbarker reviewed (changes requested):
Need the scope here or some poor person in the future gets an error... ("jump to case label" or something)

peterbarker commented:
This looks well-formed as a PR now - thanks for reworking it!

mustafa-gokce commented:
@peterbarker, thanks for your guidance.
I started working to support this command in GUIDED mode and other vehicles.
And when it is done, I will post it.
Have a nice day.

rmackay9 commented:
@mustafa-gokce,

Can you test what happens in these situations?

**Situation 1**
1. missiion is stopped using the new command
2. user switches out of auto and then back into auto
expected result: mission should restart

**Situation 2**
1. missiion is stopped using the new command
2. user changes active waypoint to some other command (e.g. "wp set 3")
3. user restarts mission with the new command
expected result: mission should restart at command 3 (?)

**Situation 3**
1. mission is stopped using the new command
2. user restarts the mission using the MISSION_START command
expected result: mission should start but I'm unsure whether it will resume or restart.  It may depend upon the WPNAV_RESTART param value.

peterbarker commented:
... I'm curious to know what happens if you include this as a mission item in a mission, @mustafa-gokce - we might need to guard against it.

peterbarker commented:
@mustafa-gokce I've added an autotest here to ensure the basic functionality never dies.

mustafa-gokce commented:
@rmackay9 hello sir.
I used this mission to test it.
```
QGC WPL 110
0	0	0	16	0.000000	0.000000	0.000000	0.000000	-35.363262	149.165237	584.049988	1
1	0	3	16	0.000000	0.000000	0.000000	0.000000	-35.359349	149.164783	100.000000	1
2	0	3	16	0.000000	0.000000	0.000000	0.000000	-35.363506	149.160206	100.000000	1
3	0	3	16	0.000000	0.000000	0.000000	0.000000	-35.367981	149.166408	100.000000	1
4	0	3	16	0.000000	0.000000	0.000000	0.000000	-35.362578	149.171537	100.000000	1
5	0	0	177	1.000000	1000.000000	0.000000	0.000000	0.000000	0.000000	0.000000	1
```
I loaded the mission and did the following:
```
mode GUIDED
arm throttle
takeoff 100
mode AUTO
```
**Situation 1**
1. Did `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0 0 0` and vehicle stopped.
2. Did `mode GUIDED` and the vehicle is still loitering.
3. Did `mode AUTO` and vehicle resuming the mission where is it left off because MIS_RESTART=0, but when set MIS_RESTART=1 and do all steps, after the 3rd step, the vehicle starts the mission from the beginning as it supposed to.
So I think it passes the first situation.

**Situation 2**
1. I loaded the mission, took off, and changed the mode to AUTO.
2. I did `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0 0 0` and vehicle holds position.
3. I did wp set 2 (while it waiting between 3 and 4).
4. I did `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 1 0 0 0 0 0 0` and vehicle starts to move 2nd wp.
So I think it passes the second situation.

**Situation 3**
1. I loaded the mission, took off, and changed the mode to AUTO.
2. I did `command_int GLOBAL MAV_CMD_DO_PAUSE_CONTINUE 0 0 0 0 0 0 0 0 0` and vehicle holds position.
3. I did `long MAV_CMD_MISSION_START 0 0 0 0 0 0 0 0` and the mission resumes where it left off because MIS_RESTART=0, but when set MIS_RESTART=1 and run it again, it starts from the beginning of the mission.
So I think it passes the third situation.

mustafa-gokce commented:
> @mustafa-gokce I've added an autotest here to ensure the basic functionality never dies.

@peterbarker I think we have a problem.
When I add command to the mission, SITL crashed, 
> PANIC: Mission command with ID 193 has no string

mustafa-gokce commented:
@peterbarker I added string for MAV_CMD_PAUSE_CONTINUE and now SITL works fine.
When I add MAV_CMD_PAUSE_CONTINUE in the auto mission list, it says it is an invalid command, skipping and resuming the mission without interrupting.

rmackay9 reviewed (approved):
@mustafa-gokce, thanks very much for doing that testing.

peterbarker reviewed (approved):
LGTM

rmackay9 commented:
Merged, thanks!

BV-OpenSource commented:
> @peterbarker I added string for MAV_CMD_PAUSE_CONTINUE and now SITL works fine. When I add MAV_CMD_PAUSE_CONTINUE in the auto mission list, it says it is an invalid command, skipping and resuming the mission without interrupting.

Should it not be working on mission mode?

I opened the discussion [here](https://discuss.ardupilot.org/t/about-the-copter-4-2-category/82459/2?u=bv-opensource) if you want to continue there.

rmackay9 commented:
@BV-OpenSource, txs for the report.  I've replied in the discussion.

mustafa-gokce commented:
@BV-OpenSource thanks for your interest, I replied in [here](https://discuss.ardupilot.org/t/about-the-copter-4-2-category/82459/5?u=mustafa_gokce)
